### PR TITLE
Improve HIR stats

### PR DESCRIPTION
**Pull request by nnethercote**
#100398 improve the AST stats collection done by `-Zhir-stats`. This PR does the same for HIR stats collection.

r? @davidtwco

### CODE REVIEW

**Comment by nnethercote** [comment]
Best reviewed one commit at a time.

File: compiler/rustc_interface/src/passes.rs [hunk 1]
```diff
@@ -416,7 +416,7 @@ pub fn configure_and_expand(
     }
 
     if sess.opts.unstable_opts.hir_stats {
-        hir_stats::print_ast_stats(&krate, "POST EXPANSION AST STATS");
+        hir_stats::print_ast_stats(&krate, "POST EXPANSION AST STATS", "ast-stats-2");
```
Comment (davidtwco) [inline compiler/rustc_interface/src/passes.rs:419]: nit: `ast-stats-2` could probably be `ast-pre-stats` or something a little more descriptive? Likewise with `ast-stats-1`.

**Review by davidtwco — APPROVED** [review]
This looks good to me, one nit, feel free to resolve it or not, r=me.

File: compiler/rustc_interface/src/passes.rs (hunk 1 above)
Comment (nnethercote) [reply compiler/rustc_interface/src/passes.rs:419]: I'll leave it as is. Anything short of `ast-stats-{pre,post}-expansion` is going to be non-obvious to a first-time user, but that's a bit on the long side. And the heading ensures that a first-time user will be able to quickly work out the meaning of the `1` and `2`.

**Pull request merged**